### DISCUSSION TITLE
Question about inconsistent code generation while using "foreach" and "List<T>"

### DISCUSSION THREAD

**Discussion by marco-carvalho** [dotnet/runtime#106817] [Q&A] [answered]
I have those methods:
```cs
static void ListForEach(List<int> numbers)
{
    foreach (int number in numbers)
    {
        Console.WriteLine(number);
    }
}

static void ListFor(List<int> numbers)
{
    for (int i = 0; i < numbers.Count; i++)
    {
        Console.WriteLine(numbers[i]);
    }
}

static void ArrayForEach(int[] numbers)
{
    foreach (int number in numbers)
    {
        Console.WriteLine(number);
    }
}

static void ArrayFor(int[] numbers)
{
    for (int i = 0; i < numbers.Length; i++)
    {
        Console.WriteLine(numbers[i]);
    }
}
```

They generate something like this:
```cs
[NullableContext(1)]
[CompilerGenerated]
internal static void <<Main>$>g__ListForEach|0_0(List<int> numbers)
{
    List<int>.Enumerator enumerator = numbers.GetEnumerator();
    try
    {
        while (enumerator.MoveNext())
        {
            Console.WriteLine(enumerator.Current);
        }
    }
    finally
    {
        ((IDisposable)enumerator).Dispose();
    }
}

[NullableContext(1)]
[CompilerGenerated]
internal static void <<Main>$>g__ListFor|0_1(List<int> numbers)
{
    int num = 0;
    while (num < numbers.Count)
    {
        Console.WriteLine(numbers[num]);
        num++;
    }
}

[NullableContext(1)]
[CompilerGenerated]
internal static void <<Main>$>g__ArrayForEach|0_2(int[] numbers)
{
    int num = 0;
    while (num < numbers.Length)
    {
        Console.WriteLine(numbers[num]);
        num++;
    }
}

[NullableContext(1)]
[CompilerGenerated]
internal static void <<Main>$>g__ArrayFor|0_3(int[] numbers)
{
    int num = 0;
    while (num < numbers.Length)
    {
        Console.WriteLine(numbers[num]);
        num++;
    }
}
```

https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEAsBuWsUDZmsQBkBLAZwBcAxAewCcBRAQwGMALAChIoB5iA7cgD5EfAK4BbAEYBTWqQCUsAN6xEaxADM60lq0Tt+5ERJm1E/Y1NkLV6lTHWPkEAJzsxV2vMwP1AX1gAmGwIPBACLio6TjJyXgFhD1MbGHtHLTMDAXNEAF5EeHQc7ktkgDoAYWpRASLiAGp6xV81NKc1FDck6wBtYgBdb1s1IKCQsIIAQVpaRgBPGgZdLPIe/tLrZra1DJ02fUMNswtuuWbHbadO9xNZIZbEUaw4UPxEadmF6MM1o5TLjIHbLEPIFOqIEqnUhlQjSPgAc3IrDqjXOdmGV1cN08pD6gx8jieMCAA=

Why only `foreach` with `List<T>` generate a different final code?

**Comment by EgorBo** [comment]
`foreach` for `List<T>` generates exactly what it is supposed to do for `foreach` over `IEnumerable`, it's just Roslyn special cases Arrays to emit better (simpler to optimize) IL.

**Comment by KalleOlaviNiemitalo** [comment, marked as answer]
`ListForEach` uses List\<T\>.Enumerator.MoveNext(), which throws InvalidOperationException if the list has been modified after the enumerator was created.

`ListFor` and `ArrayFor` don't check that because they use just an integer index rather than a enumerator object.

`ArrayForEach` doesn't check that because arrays don't keep track of modifications.

**Reply by huoyaoyuan** [reply]
Here "modification" means changing the size of the containing list. Size of array can't be changed.

**Reply by KalleOlaviNiemitalo** [reply]
The List\<T\>.this[int index] setter increments `_version` and thus counts as a modification, even though it doesn't change the size of the list.  <https://github.com/dotnet/runtime/blob/10107d3ca202bf1fda76a1bf575d782be4be27c3/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L155-L163>
